### PR TITLE
Make torchgeometry jit compatible

### PR DESCRIPTION
**Pull request by Wizaron**
In order to enable JIT tracing, following modifications are made.

1. `torchgeometry/conversions.py` : Bit length in `convert_points_to_homogeneous` is calculated using torch methods. (if we trace the model using JIT, `points.shape[-1]` is of type `torch.Tensor` instead of `int`.)
2. `torchgeometry/homography_warper.py` : `padding_mode` argument is moved from `forward` method to initialization of `HomographyWarper` since  JIT does not allow strings as arguments in `forward` methods.
3. `torchgeometry/imgwarp.py` : 
    * Tensor in `normal_transform_pixel` is created using arithmetic operations instead of replacement.
    * Generation of a 3x3 transformation matrix from 2x3 affine in `dst_norm_to_dst_norm` is updated in order to get rid of replacement of `M` to `M_3x3`.
4. `test/test_jit_tracing.py` is added in order to test JIT tracing.

### CODE REVIEW

File: torchgeometry/imgwarp.py [hunk 1]
```diff
@@ -135,9 +142,9 @@ def warp_affine(src, M, dsize, flags='bilinear', border_mode=None,
         raise ValueError("Input M must be a Bx2x3 tensor. Got {}"
                          .format(src.shape))
     # we generate a 3x3 transformation matrix from 2x3 affine
-    M_3x3 = torch.eye(3, device=src.device, dtype=src.dtype)
-    M_3x3 = torch.unsqueeze(M_3x3, dim=0).repeat(src.shape[0], 1, 1)
-    M_3x3[..., :2, :3] = M
+    M_3x3 = F.pad(M, (0, 0, 0, 1, 0, 0))
```
Comment (edgarriba) [inline torchgeometry/imgwarp.py]: sweet! :D

File: test/test_jit_tracing.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
+import pytest
+from common import TEST_DEVICES
+
+import torch
+import torchgeometry as tgm
+import torch.nn as nn
+
+
+class WarpPerspective(nn.Module):
+
```
Comment (edgarriba) [inline test/test_jit_tracing.py]: blank lines only between signatures

File: test/test_jit_tracing.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        if self.affine:
+            self.linear = nn.Linear(32 * 16 * 16, 6)
+            self.warper = WarpAffine(output_size)
+        else:
+            self.linear = nn.Linear(32 * 16 * 16, 9)
+            self.warper = WarpPerspective(output_size)
+
+    def forward(self, x):
+        bs = x.size(0)
```
Comment (edgarriba) [inline test/test_jit_tracing.py]: use meaningful and readable variable names: `bs -> batch_size`

File: test/test_jit_tracing.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+        return x
+
+
+class LocalizationNetwork(nn.Module):
+
+    def __init__(self, output_size, pretrained=True, affine=True):
+        super(LocalizationNetwork, self).__init__()
+
+        self.cnn = nn.Sequential(
```
Comment (edgarriba) [inline test/test_jit_tracing.py]: do we really need that many layers ? could just work let's say one block of conv+relu+max-pool with very few neurons (e.g. 1-3) ?

File: torchgeometry/imgwarp.py [hunk 5]
```diff
@@ -14,11 +15,17 @@
 
 
 def normal_transform_pixel(height, width):
-    return torch.tensor([[
-        [2. / (width - 1), 0., -1.],
-        [0., 2. / (height - 1), -1.],
-        [0., 0., 1.]]
-    ])  # 1x3x3
+
```
[the remark this answers is not in the document]
Comment (Wizaron) [reply torchgeometry/imgwarp.py]: JIT returns the `height` and `width` as a one element tensor (instead of integer) and it cannot replace these values inside another tensor. It simply returns

```
[[0, 0, -1]
[0, 0, -1]
[0, 0, 1]]
```
as output. Hence, I constructed the tensor using arithmetic operations instead of replacing them into another tensor.

File: torchgeometry/imgwarp.py (hunk 1 above)
Comment (Wizaron) [reply torchgeometry/imgwarp.py]: Again, JIT cannot replace a tensor into another. If we want to replace `M` into `M_3x3` (instead of constructing it with padding and an addition), it throws the following error which states that outputs of pytorch and torch scripts models are not equivalent:

```
  /home/jcmaxwell/miniconda2/envs/pytorch-1.0-stable/lib/python3.7/site-packages/torch/jit/__init__.py:642: TracerWarning: Output nr 1. of the traced function does not match the corresponding output of the Python function. Detailed error:
  Not within tolerance rtol=1e-05 atol=1e-05 at input[0, 0, 50, 34] (4.5528998374938965 vs. 0.0) and 49150 other locations (99.00%)
    _check_trace([example_inputs], func, executor_options, module, check_tolerance, _force_outplace)
```

This error directs us a simpler solution :+1: Thanks!

File: test/test_jit_tracing.py (hunk 4 above)
Comment (Wizaron) [reply test/test_jit_tracing.py]: I have just simplified the model.

File: test/test_jit_tracing.py (hunk 3 above)
Comment (Wizaron) [reply test/test_jit_tracing.py]: Agreed :+1:

File: torchgeometry/conversions.py [hunk 6]
```diff
@@ -115,8 +115,16 @@ def convert_points_to_homogeneous(points):
         raise ValueError("Input must be at least a 2D tensor. Got {}".format(
             points.shape))
 
+    if torch.is_tensor(points.shape[-1]):
+        last_dim_shape = points.shape[-1].float()
+        last_dim_bit_length = torch.floor(torch.log2(last_dim_shape)) + 1
+        last_dim_bit_length = last_dim_bit_length.long()
```
[the remark this answers is not in the document]
Comment (edgarriba) [reply torchgeometry/conversions.py]: gotcha. However, I think we are over complicating things, similar to your other proposal with just padding the tensor it should work right ? `torch.nn.functional.pad(x, (0, 1), "constant", 1.0)`

File: torchgeometry/imgwarp.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (edgarriba) [reply torchgeometry/imgwarp.py]: I see. I think there is a clean alternative so that changes are minimal. If this two guys** are converted to `torch.tensor` and we assume that `normal_transform_pixel` consumes just tensor (need a type check there) I guess then that we solve the issue.

**https://github.com/arraiy/torchgeometry/blob/master/torchgeometry/imgwarp.py#L26-L27

File: torchgeometry/conversions.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Wizaron) [reply torchgeometry/conversions.py]: We can definitely pad the tensor, instead of concatenating it with ones. However, we still need to calculate bit length of the last dimension in order to determine the size of the padding.

[the remark this answers is not in the document]
Comment (edgarriba) [reply torchgeometry/conversions.py]: even another option to avoid bit length compute would be something like that:
`return torch.cat([points, torch.ones_like(points)[..., :1]], dim=-1)`

we are allocating unnecessary memory but I don't think I will harm the performance at all and it's clean.

[the remark this answers is not in the document]
Comment (edgarriba) [reply torchgeometry/conversions.py]: or wait, that one might work better: D
`return torch.cat([points, torch.ones_like(points)[..., -1:]], dim=-1)`

[the remark this answers is not in the document]
Comment (edgarriba) [reply torchgeometry/conversions.py]: anyways, whatever approach you choose please provide a unittest for JIT

[the remark this answers is not in the document]
Comment (Wizaron) [reply torchgeometry/conversions.py]: I have made a commit that removes bit length computation and constructs the tensor via padding. I will try to provide JIT versions of the unit tests.

**Comment by edgarriba** [comment]
@Wizaron are you in the pytorch slack workspace ? we just opened a channel called `#geometry` to discuss stuff about torchgeometry. Would you like to join ?
